src/include/denc.h

```diff
@@ -0,0 +1,1176 @@
+  is encoded.
+
+  - denc_traits<T> are declared for the base integer types, string, bufferptr,
+  and bufferlist base types.
+
+  - denc_traits<std::foo<T>>-like traits are declared for standard container
+  types.
+
+
+  class methods look like so
+  ==========================
+
+    void bound_encode(size_t& p) const;
+    void encode(buffer::list::contiguous_appender& p) const;
+    void decode(buffer::ptr::iterator &p);
+
+  or (for featured objects)
+
+    void bound_encode(size_t& p, uint64_t f) const;
+    void encode(buffer::list::contiguous_appender& p, uint64_t f) const;
+    void decode(buffer::ptr::iterator &p);
+
+  - These are normally invoked by the denc_traits<> methods that are
+  declared via WRITE_CLASS_DENC, although you can also invoke them explicitly
+  in your code.
+
+  - These can be defined either explicitly (as above), or can be "magically"
+  defined all in one go using the DENC macro and DENC_{START,FINISH} helpers
+  (which work like the legacy {ENCODE,DECODE}_{START,FINISH} macros):
+
+    class foo_t {
+      ...
+      DENC(foo_t, v, p) {
+        DENC_START(1, 1, p);
+        denc(v.foo, p);
+        denc(v.bar, p);
+        denc(v.baz, p);
+        DENC_FINISH(p);
+      }
+      ...
+    };
+    WRITE_CLASS_DENC(foo_t)
+
+  */
+
+
+// -----------------------------------------------------------------------
+// base types
+
+// Just like the existing encode/decode machinery. The environment
+// provides a rich set of pre-defined encodes for primitive types and
+// containers
+
+// These macros all handle byte-ordering automatically for you :) All
+// external serializations are in little endian order
+
+// itype == internal type
+// otype == external type, i.e., the type on the wire
+
+// NOTE: set supported == 2 instead of true.  This prevents these from
+// getting glued into the legacy encode/decode methods; the overhead
+// of setting up a contiguous_appender etc is likely to be slower.
+
+#define DEFINE_DENC_SIMPLE(itype, etype)				\
+  template<>								\
+  struct denc_traits<itype> {						\
+    enum { supported = 2 };						\
+    enum { featured = false };						\
+    enum { bounded = true };						\
+    static void bound_encode(const itype &o, size_t& p, uint64_t f=0) {	\
+      p += sizeof(etype);						\
+    }									\
+    static void encode(const itype &o, buffer::list::contiguous_appender& p, \
+		       uint64_t f=0) {					\
+      *(etype *)p.get_ptr_add(sizeof(etype)) = static_cast<etype>(o);	\
+    }									\
+    static void decode(itype& o, buffer::ptr::iterator &p) {		\
+      o = *(etype*)p.c_str_add(sizeof(etype));				\
+    }									\
+  };
+
+DEFINE_DENC_SIMPLE(uint8_t, uint8_t);
+DEFINE_DENC_SIMPLE(int8_t, int8_t);
+DEFINE_DENC_SIMPLE(uint16_t, __le16);
+DEFINE_DENC_SIMPLE(int16_t, __le16);
+DEFINE_DENC_SIMPLE(uint32_t, __le32);
+DEFINE_DENC_SIMPLE(int32_t, __le32);
+DEFINE_DENC_SIMPLE(uint64_t, __le64);
+DEFINE_DENC_SIMPLE(int64_t, __le64);
+
+
+// varint
+//
```

Comment:
You can use the template mechanism to automatically select between signed and unsigned varint encoding. It'll simplify the namespace a bit....